In [1]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_1samp, chisquare

In [2]:
# Extracting the data from the file
talent_df = pd.read_csv('talent_survey - talent_survey.csv')

talent_df.head(10)

,Token,Submitted At,Respondent gender,Age Group,Institution category,Please rate your satisfaction with the quality of the lectures you receive,I think the course is designed in a way that meets the current tech industry needs,How many separate industrial trainings are you expected to complete in your department?,What programming languages were you exposed to in the school curriculum?
0,3iecc17hcyssjxsvos923iecc17fg2zb,2023-01-26T09:33:23Z,Male,20 to 24,University,4,NaN,NaN,NaN
1,79cmx9f35a8rkpzg5yyvz79cmlq1b7ma,2023-02-09T20:43:25Z,Male,20 to 24,Polytechnic,4,2.0,2.0,"FORTRAN, JavaScript, Java, C++, Python, C#, C,..."
2,guvgwczzamuzfti9rguv2g2j7qso42c0,2023-01-10T18:32:38Z,Male,20 to 24,University,3,2.0,1.0,"Java, Python"
3,9d6kacm78oea59bd9dyglv2ykut38xku,2023-01-26T18:00:00Z,Male,20 to 24,University,3,NaN,NaN,NaN
4,rjnc2lqphru2g0nt3bvtirjnc97xevrr,2023-01-23T17:39:49Z,Male,20 to 24,University,1,1.0,1.0,NaN
5,yflowz4s55u1r9iyflx4woxzh0jge6j8,2023-01-25T06:20:43Z,Male,20 to 24,University,3,3.0,2.0,NaN
6,dwb49tji9e0en1o9ipxgnzsdwb4q85a7,2023-01-24T22:20:15Z,Female,20 to 24,University,4,3.0,2.0,"JavaScript, Java, Python, C++, C#, BASIC, FORTRAN"
7,a56jfgmrcgm2p5q6nz04a56ju1bpv2r1,2023-02-09T20:50:29Z,Female,20 to 24,Polytechnic,3,NaN,NaN,"SQL, C, BASIC, FORTRAN, Java"
8,7a75h6zbge78lfuyzbbi7a75hm7on2e2,2023-02-10T08:15:01Z,Male,20 to 24,Polytechnic,5,4.0,1.0,"Python, C#, FORTRAN, BASIC"
9,pn9ngu4onjspzlk1tz6pn9hunaz5m24k,2023-01-27T05:18:06Z,Male,20 to 24,University,4,NaN,1.0,NaN


In [3]:
talent_df.shape

(250, 9)

In [4]:
talent_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 9 columns):
 #   Column                                                                                   Non-Null Count  Dtype  
---  ------                                                                                   --------------  -----  
 0   Token                                                                                    250 non-null    object 
 1   Submitted At                                                                             250 non-null    object 
 2   Respondent gender                                                                        250 non-null    object 
 3   Age Group                                                                                250 non-null    object 
 4   Institution category                                                                     250 non-null    object 
 5   Please rate your satisfaction with the quality of the lectures yo

#### Generate descriptive statistics for the demographic characteristics of the respondents.

In [5]:
talent_df.describe()

,Please rate your satisfaction with the quality of the lectures you receive,I think the course is designed in a way that meets the current tech industry needs,How many separate industrial trainings are you expected to complete in your department?
count,250.000000,184.000000,165.000000
mean,3.056000,2.798913,1.624242
std,1.066496,1.186352,1.008207
min,1.000000,1.000000,0.000000
25%,3.000000,2.000000,1.000000
50%,3.000000,3.000000,1.000000
75%,4.000000,4.000000,2.000000
max,5.000000,5.000000,6.000000


#### Identify the most common programming languages that students were exposed to during their coursework.

In [6]:
talent_df.columns

Index(['Token', 'Submitted At', 'Respondent gender', 'Age Group',
       'Institution category',
       'Please rate your satisfaction with the quality of the lectures you receive',
       'I think the course is designed in a way that meets the current tech industry needs',
       'How many separate industrial trainings are you expected to complete in your department?',
       'What programming languages were you exposed to in the school curriculum?'],
      dtype='object')

In [7]:
talent_df = talent_df.rename(columns={'What programming languages were you exposed to in the school curriculum?':'Programming languages'})

# Splitting the strings in the programming languages into strings.

talent_df['Programming languages split'] = talent_df['Programming languages'].str.split(', ')
split_col = talent_df['Programming languages split'].dropna()

# Flatten the list of lists into a single flat list
import itertools
languages = list(itertools.chain.from_iterable(split_col))

common_prog_languages = pd.Series(languages).value_counts()

# Getting the top 5 programming languages

common_prog_languages.head(5)

C++           56
Python        54
Java          48
BASIC         47
JavaScript    34
Name: count, dtype: int64

#### Use an appropriate statistical method to test whether the average number of required 
indusrial  training sessions is equal to 2.

In [8]:
# Null hypothesis: mean = 2
# Alternative hypothesis: mean is not 2

# Using the one-sample t-test we will compare the actual mean with the mean given above.
from scipy.stats import ttest_1samp

talent_df = talent_df.rename(columns={'How many separate industrial trainings are you expected to complete in your department?': 'Trainings'})
trainings = talent_df['Trainings'].dropna()

hyp_value = 2
t_statistic, p_value = ttest_1samp(trainings, hyp_value)
p_value
# print(f"T-statistic: {t_statistic:.2f}")
# print(f"P-value: {p_value:.3f}")

3.750598062741265e-06

A one-sample t-test was performed to evaluate whether the average number of require industrial trainings is 2.
The The test returned a p-value of 0.00000375 which is extremely below 0.05 significance level.
This provides strong evidence to reject the null hypothsis, suggesting that the average number of required trainings is significantly different from 2.

#### Apply a suitable inferential statistics method to draw conclusions about student ssatisfactio  with the quality of lectures received.

Where 1 = Very dissatisfied, 2 = Dissatisfied, 3 = Indifferent, 4 = Satisfied, 5 = Very
satisfied

I would like to test whether avergae satisfaction differs from a neutral point which is indeffernce
i.e 3 = Indifferent

Using the one-sample t-test
Null hypothesis: The average satisfaction score is 3 (indifference)
Alternative hypothesis: The average satisfaction is not 3 (so students are either satisfied or dissatisfied)



In [9]:
talent_df = talent_df.rename(columns={'Please rate your satisfaction with the quality of the lectures you receive': 'Satisfaction'})
# talent_df.head()

hyp_value1 = 3
t_stat, p_val = ttest_1samp(talent_df['Satisfaction'], hyp_value1)

print(f"T-statistic: {t_stat:.3f}")
print(f"P-value: {p_val:.3f}")

T-statistic: 0.830
P-value: 0.407


The t-statistic is 0.830, indicating that the sample mean is only 0.83 standard errors away from the hypothesized mean of 3.
The p-value is 0.407, which is greater than the conventional threshold of 0.05. This means the difference is not statistically significant.
Therefore, we fail to reject the null hypothesis — students' satisfaction levels do not significantly differ from indifference.

#### Use the chi-square method to analyze students' perceptions of whether the course design meets the current needs of the tech industry.

where 1=strongly disagree, 2=disagree, 3=indifferent, 4=agree, and 5=strongly agree

Choosing the null hypothesis to be:

Null hypothesis: Students are equally likely to agree or disagree.

Alternative hypothesis: Students are not equally likely to agree or disagree.


We will have two groups of responses:

Agree: 3=indifferent, 4=agree, and 5=strongly agree

Disagree: 1=strongly disagree, 2=disagre

In [16]:
talent_df = talent_df.rename(columns={'I think the course is designed in a way that meets the current tech industry needs': 'Course_perception'})
filtered_df = talent_df.dropna(subset=['Course_perception']) # dropping the missing values so as not to skew the results

disagree = [1,2]
agree = [3,4,5]

filtered_df['Perception'] = np.where(filtered_df['Course_perception'].isin(agree), 'Agree', 'Disagree')
# filtered_df.head()

# To perform the chi-squared test
observed_values = filtered_df['Perception'].value_counts()
observed_values # Agree = 115%, Disagree = 69

total = observed_values.sum()
total

# Going by the null hypothesis, then 50% should be disagreeing and 50% agree 
expected_values = [0.5 * total, 0.5 * total]

observation = [observed_values['Disagree'], observed_values['Agree']]

chi2_statistic, p_value = chisquare(f_obs=observation, f_exp=expected_values)
print(f"Chi-square: {chi2_statistic:.3f}")
print(f"P-value: {p_value:.4f}")

Chi-square: 11.500
P-value: 0.0007


C:\Users\RMwaura\AppData\Local\Temp\ipykernel_22556\4051528899.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['Perception'] = np.where(filtered_df['Course_perception'].isin(agree), 'Agree', 'Disagree')


Course_perception
3.0    60
4.0    43
1.0    35
2.0    34
5.0    12
Name: count, dtype: int64

Chi-square: 11.500

P-value: 0.0007

Since the p-value < 0.05, the result is statistically significant, therefore, we reject the null hypothesis that student perceptions of the course design are equally split (50/50) between agreement and disagreement.

This suggests that students are more likely to either agree or disagree that the course design meets the current needs of the tech industry.